web/src/narrow_state.ts

```diff
@@ -265,6 +268,17 @@ export function _possible_unread_message_ids(
     let topic_name;
     let current_filter_pm_string;
 
+    if (current_filter.can_bucket_by("channel", "topic", "with")) {
+        const with_operand = current_filter.operands("with")[0];
+        assert(with_operand !== undefined);
+        const msg_id = Number.parseInt(with_operand, 10);
+        const msg = message_store.get(msg_id);
+        if (msg !== undefined && msg.type === "stream") {
+            return unread.get_msg_ids_for_topic(msg.stream_id, msg.topic);
+        }
+        return [];
+    }
```

Comment:
So this could reasonably be a follow-up PR, but if we have the target message locally, we should be able to have `message_view.ts` resolve the correct channel/topic to use and set `filter.requires_adjustment_for_moved_with_target` appropriately without talking to the server, which seems like a very common and useful optimization that would apply to the vast majority of these links.

Reply:
The above suggestion was implemented. But I'm left with the question of whether this logic is correct -- it seems like if the target message is not present in `message_store`, then it'll return `[]`, and I'm not sure whether callers will do something sane in that scenario -- unlike other cases in this function, this is not a case where we know there are 0 unreads in the conversation, but instead one where we don't know.

It might be most appropriate to raise an exception in the case that the message is not in the message store, here? 
But regardless we need to audit how this is used to make sure it's safe, and have a plan to ensure this doesn't break in future refactoring..

Or maybe assert `this.requires_adjustment_for_moved_with_target` and just use the channel/topic, so we know we've got the right values to inspect?

Reply:
It does seem safe, because this is only called inside `maybe_add_local_messages`, which we skip if `!filter.requires_adjustment_for_moved_with_target`. So probably this can be addressed with an assert and clear comment.

Reply:
I made these changes, which I think make this section of code obviously correct:

``` diff
diff --git a/web/src/narrow_state.ts b/web/src/narrow_state.ts
index 7d9e79a61c..b666739a66 100644
--- a/web/src/narrow_state.ts
+++ b/web/src/narrow_state.ts
@@ -268,15 +268,27 @@ export function _possible_unread_message_ids(
     let topic_name;
     let current_filter_pm_string;
 
+    // For the `with` operator, we can only correctly compute the
+    // correct channel/topic for lookup unreads in if we either
+    // have the message in our local cache, or we know the filter
+    // has already been updated for potentially moved messages.
+    //
+    // The code path that needs this function is never called in
+    // the `with` code path, but for safety, we assert that
+    // assumption is not violated.
+    //
+    // If we need to change that assumption, we can try looking up the
+    // target message in message_store, but would need to return
+    // undefined if the target message is not available.
+    assert(!current_filter.requires_adjustment_for_moved_with_target);
+
     if (current_filter.can_bucket_by("channel", "topic", "with")) {
-        const with_operand = current_filter.operands("with")[0];
-        assert(with_operand !== undefined);
-        const msg_id = Number.parseInt(with_operand, 10);
-        const msg = message_store.get(msg_id);
-        if (msg !== undefined && msg.type === "stream") {
-            return unread.get_msg_ids_for_topic(msg.stream_id, msg.topic);
+        sub = stream_sub(current_filter);
+        topic_name = topic(current_filter);
+        if (sub === undefined || topic_name === undefined) {
+            return [];
         }
-        return [];
+        return unread.get_msg_ids_for_topic(sub.stream_id, topic_name);
     }
 
     if (current_filter.can_bucket_by("channel", "topic")) {
```